### PR TITLE
make c# api conf more flexible when config internal log

### PR DESCRIPTION
**Pull request by lust4life**
api config for c# is only support console target for logging internal logs right now , can add flexible for the case use asp.net in vs, can use debugger or some other targets for internal logging.

```csharp
LogaryFactory.New("demoService",
                    conf => conf
                            .InternalLoggingLevel(LogLevel.Verbose)                        
                            .Target<Debugger.Builder>("debugger", tb => tb.UseForInternalLog())
                            .Target<Logary.Targets.Console.Builder>("console", tb => tb.UseForInternalLog())
                            .Target<LiterateConsole.Builder>("console1")                            
                            .Target<AliYun.Builder>("AliYunLog", BuildAliYunTarget)
                    );
```

- [x] add sample code in readme

- [ ] add some test

```fsharp
    // generate msg use default messagetemplates
    let formatedMsg = Formatting.format (Parser.parse formatTemplate) args
```
csharp LoggerExtensions when log event with object args ,  when use fsharp , there is no chance to add othere type in `fields    : Map<PointName, Field>` ， but when use in csharp  with the `LogEventFormat` method , we use `args : obj[]` for templete . 

use `MessageTemplate` as default format, when invoke `LogEventFormat` from c# api.

### CODE REVIEW

File: src/Logary/Configuration_FactoryApi.fs [hunk 1]
```diff
@@ -57,14 +65,51 @@ and internal MetricsConfBuilder(conf) =
...
+  let configInternalTargets conf =
+    // find all targets set for internal log through a specific rule,
+    // and remove these (specific rule/target) from logary conf, so normal logging is not affected. 
+    // if there isn't any, use console as default
+
+    let (normalRules, forInternalRules) =
+      conf.rules 
+      |> List.partition (fun r -> 
+            match r with
+            | r when r.hiera.ToString() = specificHieraForSetInternalLog -> false
```
Comment (haf) [inline src/Logary/Configuration_FactoryApi.fs]: A hiera is a regex and should be used as such.

File: src/Logary/Configuration_FactoryApi.fs [hunk 2]
```diff
@@ -57,14 +65,51 @@ and internal MetricsConfBuilder(conf) =
 
 /// The "main" fluent-config-api type with extension method for configuring
 /// Logary rules as well as configuring specific targets.
-and ConfBuilder(conf) =
+and ConfBuilder(conf, internalLoggingLevel) =
+
+  let specificHieraForSetInternalLog = "^specificHieraForSetInternalLog$"
```
Comment (haf) [inline src/Logary/Configuration_FactoryApi.fs]: Why can't we represent this case in types?

File: src/Logary/Configuration_FactoryApi.fs [hunk 3]
```diff
@@ -162,5 +214,5 @@ type LogaryFactory =
     if serviceName = null then nullArg "serviceName"
     if configurator = null then nullArg "configurator"
     let config = Config.confLogary serviceName
-    let cb = configurator.Invoke (ConfBuilder config)
-    cb.BuildLogary () |> CSharp.toTask
\ No newline at end of file
+    let cb = configurator.Invoke (ConfBuilder (config,LogLevel.Fatal) )
```
Comment (haf) [inline src/Logary/Configuration_FactoryApi.fs]: Where's this level used?

File: src/Logary/Configuration_FactoryApi.fs [hunk 4]
```diff
@@ -128,10 +175,15 @@ and ConfBuilder(conf) =
     // referentially transparent
     let targetConf = f.Invoke(!contRef) :?> ConfBuilderT<'T>
 
+    let tr = 
+      match targetConf.useForInternalLog with
```
Comment (haf) [inline src/Logary/Configuration_FactoryApi.fs]: Prefer if-s

File: src/Logary/Configuration_FactoryApi.fs [hunk 5]
```diff
@@ -57,14 +65,51 @@ and internal MetricsConfBuilder(conf) =
...
+          forInternalRules 
+          |> List.exists (fun r -> r.target = k.ToString())
+          |> not)
+
+    let internalTargets = 
+      internalTargets 
+      |> Map.toList
+      |> List.map (fun (_,tci) -> fst tci) 
+      |> function 
+        | [] -> [Console.create Console.empty "internalConsole"]
```
Comment (haf) [inline src/Logary/Configuration_FactoryApi.fs]: Can't we default to this in the building?

Comment (lust4life) [inline src/Logary/Configuration_FactoryApi.fs]: it is the default , see comments after.

File: src/Logary/Configuration_FactoryApi.fs (hunk 3 above)
Comment (lust4life) [inline src/Logary/Configuration_FactoryApi.fs]: it is the default internal logging level , use when people not setting it from `InternalLoggingLevel` . see comments after.

File: src/Logary/Configuration_FactoryApi.fs (hunk 1 above)
Comment (lust4life) [inline src/Logary/Configuration_FactoryApi.fs]: here is a trick, in order to identity which targets are set for internal logging.  the specific rule with hiera `specificHieraForSetInternalLog ` is for matching a target through its `rule.target` property. 

not find other good way to do it , maybe we can use some conventions to do it , like:  a target with name match some  hiera is the internal target . but maybe affect some exists code ?

File: src/Logary/Configuration_FactoryApi.fs (hunk 2 above)
Comment (lust4life) [inline src/Logary/Configuration_FactoryApi.fs]: more details ? did not get it ☹

File: src/Logary/Configuration_FactoryApi.fs (hunk 1 above)
Comment (haf) [inline src/Logary/Configuration_FactoryApi.fs]: Seems like a boolean

**Comment by haf** [comment]
+1

File: src/Logary/Configuration_FactoryApi.fs (hunk 4 above)
Comment (lust4life) [inline src/Logary/Configuration_FactoryApi.fs]: done.

**Comment by haf** [comment]
WIP or ready?

**Comment by lust4life** [comment]
ready.

**Comment by haf** [comment]
Thank you @lust4life !

**Pull request merged**